applications/serial_lte_modem/src/slm_at_httpc.c

```diff
@@ -0,0 +1,736 @@
+				ssize_t ret;
+
+				ret = send(sock, data_buf + httpc.pl_sent,
+					   MIN(httpc.pl_to_send - httpc.pl_sent,
+					   NET_IPV4_MTU), 0);
+				if (ret < 0) {
+					LOG_ERR("send fail: %d", ret);
+					httpc.pl_len = 0;
+					httpc.pl_sent = 0;
+					k_sem_give(&http_data_sem);
+					return -errno;
+				}
+				LOG_DBG("send %d bytes payload", total_sent);
+				httpc.pl_sent += ret;
+				total_sent += ret;
+			}
+			if (total_sent == httpc.pl_len) {
+				httpc.pl_len = 0;
+			}
+			k_sem_give(&http_data_sem);
+		} while (total_sent < httpc.pl_len);
+		sprintf(rsp_buf, "#XHTTPCREQ:0\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	} else {
+		sprintf(rsp_buf, "#XHTTPCREQ:0\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	}
+
+	return total_sent;
+}
+
+static int do_http_connect(void)
+{
+	/* Connect to server if it is not connected yet. */
+	if (httpc.fd == INVALID_SOCKET) {
+		httpc.fd = server_connect(httpc.host, httpc.sec_tag);
+		if (httpc.fd < 0) {
+			LOG_ERR("server_connect fail.");
+			httpc.fd = INVALID_SOCKET;
+		}
+	}
+	sprintf(rsp_buf, "#XHTTPCCON:1\r\n");
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	return httpc.fd;
+}
+
+static int do_http_disconnect(void)
+{
+	int err = -EINVAL;
+
+	/* Close socket if it is connected. */
+	if (httpc.fd != INVALID_SOCKET) {
+		close(httpc.fd);
+		httpc.fd = INVALID_SOCKET;
+		err = 0;
+	} else {
+		return -ENOTCONN;
+	}
+	if (httpc.pl_len > 0) {
+		LOG_ERR("Exit request");
+		httpc.pl_len = 0;
+		k_sem_give(&http_req_sem);
+	}
+	sprintf(rsp_buf, "#XHTTPCCON:0\r\n");
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	return err;
+}
+
+static int http_method_str_enum(uint8_t *method_str)
+{
+	static const char * const method_strings[] = {
+		"DELETE", "GET", "HEAD", "POST", "PUT", "CONNECT", "OPTIONS",
+		"TRACE", "COPY", "LOCK", "MKCOL", "MOVE", "PROPFIND",
+		"PROPPATCH", "SEARCH", "UNLOCK", "BIND", "REBIND", "UNBIND",
+		"ACL", "REPORT", "MKACTIVITY", "CHECKOUT", "MERGE", "M-SEARCH",
+		"NOTIFY", "SUBSCRIBE", "UNSUBSCRIBE", "PATCH", "PURGE",
+		"MKCALENDAR", "LINK", "UNLINK"};
+
+	for (int i = HTTP_DELETE; i <= HTTP_UNLINK; i++) {
+		if (!strncmp(method_str, method_strings[i],
+			HTTPC_METHOD_LEN)) {
+			return i;
+		}
+	}
+	return -1;
+}
+
+static int do_http_request(void)
+{
+	int err = -EINVAL;
+	struct http_request req;
+	int method;
+	int32_t timeout = 0;
+
+	if (httpc.fd == INVALID_SOCKET) {
+		LOG_ERR("Remote host is not connected.");
+		return -EINVAL;
+	}
+
+	method = http_method_str_enum(httpc.method_str);
+	if (method < 0) {
+		LOG_ERR("Request method is not allowed.");
+		return -EINVAL;
+	}
+
+	memset(&req, 0, sizeof(req));
+	req.method = (enum http_method)method;
+	req.url = httpc.resource;
+	req.host = httpc.host;
+	req.protocol = "HTTP/1.1";
+	req.response = response_cb;
+	req.recv_buf = data_buf;
+	req.recv_buf_len = HTTPC_BUF_LEN;
+	req.payload_cb =  payload_cb;
+	req.optional_headers_cb = headers_cb;
+	err = http_client_req(httpc.fd, &req, timeout, "");
```

Comment:
Does it work like this? It kinds of contradicts with the `http_client_req` API documentation:
```
 * @param timeout Max timeout to wait for the data. The timeout value cannot be
 *        0 as there would be no time to receive the data.
 *        The timeout value is in milliseconds.
```
I guess it would be safer to set some actual value here, other than 0.

Reply:
The intention in original code is to disable timeout of http request. The socket timeout was enabled socket_timeout_set().

The zero timeout is now fixed by changing the timeout from 0 to SYS_FOREVER_MS to match the correct usage of API.